apps/studio/src/features/editing-experience/components/form-builder/renderers/controls/JsonFormsEmbedControl.tsx

```diff
@@ -0,0 +1,242 @@
+
+import { JSON_FORMS_RANKING } from "~/constants/formBuilder"
+import { useZodForm } from "~/lib/form"
+import {
+  EMBED_NAME_MAPPING,
+  getEmbedNameFromUrl,
+  getIframeSrc,
+} from "../../../utils"
+
+const SUPPORTED_MAPS = Object.keys(MAPS_EMBED_URL_REGEXES).map(
+  (key) => EMBED_NAME_MAPPING[key as keyof typeof MAPS_EMBED_URL_REGEXES],
+)
+
+const SUPPORTED_VIDEOS = Object.keys(VIDEO_EMBED_URL_REGEXES).map(
+  (key) => EMBED_NAME_MAPPING[key as keyof typeof VIDEO_EMBED_URL_REGEXES],
+)
+
+export const jsonFormsEmbedControlTester: RankedTester = rankWith(
+  JSON_FORMS_RANKING.TextControl,
+  and(
+    isStringControl,
+    schemaMatches((schema) => schema.format === "embed"),
+  ),
+)
+
+interface EmbedCodeModalProps {
+  isOpen: boolean
+  onClose: () => void
+  onSave: (embedCode: string) => void
+  urlPattern?: string
+}
+
+function EmbedCodeModal({
+  isOpen,
+  onClose,
+  onSave,
+  urlPattern,
+}: EmbedCodeModalProps) {
+  const {
+    register,
+    handleSubmit,
+    reset,
+    formState: { errors, isValid },
+  } = useZodForm({
+    mode: "onChange",
+    schema: z.object({
+      embedCode: z
+        .string()
+        .min(1, "Embed code is required")
+        .refine(
+          (value) => {
+            const iframeSrc = getIframeSrc(value)
+
+            if (!urlPattern || !iframeSrc) {
+              return !!iframeSrc
+            }
+
+            return new RegExp(urlPattern).test(iframeSrc)
+          },
+          {
+            message:
+              "This code doesn’t look valid. Copy and paste the code as-is without modifications.",
+          },
+        ),
+    }),
+    reValidateMode: "onChange",
+  })
+
+  const onSubmit = handleSubmit(({ embedCode }) => {
+    onClose()
+    onSave(embedCode)
+    reset()
+  })
+
+  return (
+    <Modal isOpen={isOpen} onClose={onClose}>
+      <ModalOverlay />
+      <ModalContent>
+        <form onSubmit={onSubmit}>
+          <ModalHeader mr="3.5rem">Insert embed code</ModalHeader>
+          <ModalCloseButton size="lg" />
+
+          <ModalBody>
+            <FormControl mt="1rem" isRequired isInvalid={!!errors.embedCode}>
+              <FormLabel description="The template may override settings like width and height of the iframe">
+                Embed code
+              </FormLabel>
+
+              <Textarea
+                placeholder="Paste embed code here"
+                fontFamily="monospace"
+                minAutosizeRows={5}
+                {...register("embedCode")}
+              />
+
+              <FormErrorMessage>{errors.embedCode?.message}</FormErrorMessage>
+            </FormControl>
+
+            <Infobox size="sm" variant="info" mt="1.25rem">
+              <Box>
+                <Text>You can embed content from:</Text>
+                <UnorderedList ml="1.5rem" mt="0.25rem">
+                  <ListItem>Video: {SUPPORTED_VIDEOS.join(", ")}</ListItem>
+                  <ListItem>Map: {SUPPORTED_MAPS.join(", ")}</ListItem>
```

Comment:
This component is shared between both maps and videos, so both lines will always appear

However, this can be confusing on the UI side because `Map: Google Map, OneMap` shows up when user is trying to add Video instead

Perhaps we can detect and only show the relevant ones?

Reply:
Hmm yeah, but mainly wanted to keep things simple first because we are not sure the universe of the embeds yet, and this is still initial designs pending improvements. I think once confirmed, we can improve this logic.